client/client.go

```diff
@@ -803,53 +780,148 @@ func (c *Client) Call(ctx context.Context, procedure string, options wamp.Dict,
-			// Let's check: was ppt feature announced by dealer?
-			if !c.sess.HasFeature(wamp.RoleDealer, wamp.FeaturePayloadPassthruMode) {
-				// It's protocol violation, so we need to abort connection
-				abortMsg := wamp.Abort{
-					Reason: wamp.ErrProtocolViolation,
-					Details: wamp.Dict{
-						wamp.OptError:   ErrPPTNotSupportedByRouter.Error(),
-						wamp.OptMessage: ErrPPTNotSupportedByPeer.Error(),
-					},
-				}
-				c.sess.Peer.Send(&abortMsg)
-				c.sess.Peer.Close()
-				return nil, fmt.Errorf("cannot process unexpected passthrough result: %w", ErrPPTNotSupportedByRouter)
-			}
+		err := c.prepareCallResultMessage(msg)
 
-			if !isPPTSchemeValid(pptScheme) {
-				return nil, fmt.Errorf("cannot process result with invalid ppt scheme %q: %w", pptScheme, ErrPPTSchemeInvalid)
-			}
+		if err != nil {
+			return nil, err
+		}
 
-			var args wamp.List
-			var kwargs wamp.Dict
-			var err error
+		return msg, nil
+	case *wamp.Error:
+		return nil, RPCError{msg, procedure}
+	default:
+		return nil, unexpectedMsgError(msg, wamp.RESULT)
+	}
+}
 
-			// Now need to check ppt_serializer (in pair with ppt_scheme)
-			// and deserialize payload with appreciate serializer
-			if pptScheme == WampPPTScheme {
-				args, kwargs, err = unpackE2EEPayload(msg.Details, msg.Arguments)
-			} else {
-				args, kwargs, err = unpackPPTPayload(msg.Details, msg.Arguments)
+// CallProgressive makes progressive calls to the procedure corresponding to
+// the given URI.
+//
+// @see full description for Call method above. The only difference is that this
+// method returns a callback that client code may call to pass a progressive chunks
+// of data to the Callee. Also, please note that all progressive results (if they
+// will be at all) will be returned using progcb callback and the final result will be
+// returned using finalrescb callback
+//
+// This function returns a callback function that should be called with payload data chunks
+// To finish a progressive call the last callback MUST be called with isLast = true
+func (c *Client) CallProgressive(ctx context.Context, procedure string, options wamp.Dict, args wamp.List, kwargs wamp.Dict, finalrescb ResultHandler, progcb ProgressHandler) (SendProgressiveData, error) { //nolint:lll
+	if !c.Connected() {
+		return nil, ErrNotConn
+	}
+
+	if !c.sess.HasFeature(wamp.RoleDealer, wamp.FeatureProgressiveCalls) {
+		return nil, ErrProgCallNotSupportedByRouter
+	}
+
+	if options == nil {
+		options = wamp.Dict{}
+	}
+
+	options[wamp.OptProgress] = true
+
+	// If caller is willing to receive progressive results, create a channel to
+	// receive these on.  Then, start a goroutine to receive progressive
+	// results and call the callback for each.
+	var progChan chan *wamp.Result
+	var progDone chan struct{}
+	if progcb != nil {
+		options[wamp.OptReceiveProgress] = true
+		progChan = make(chan *wamp.Result)
+		progDone = make(chan struct{})
+		go func() {
```

Comment:
What happens if the context gets cancelled? Will this naturally terminate because `progChan` gets closed?

Reply:
Yep. If I understand correctly, later after `msg, err := c.waitForReplyWithCancel(ctx, id, procedure, progChan)` if chan was opened — it is closed then